sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala

```diff
@@ -104,17 +155,30 @@ trait AnalysisHelper extends QueryPlan[LogicalPlan] { self: LogicalPlan =>
 
   /** Similar to [[resolveOperatorsUp]], but does it top-down. */
   def resolveOperatorsDown(rule: PartialFunction[LogicalPlan, LogicalPlan]): LogicalPlan = {
-    if (!analyzed) {
+    resolveOperatorsDownWithPruning(AlwaysProcess.fn, UnknownRuleId)(rule)
+  }
+
+  /** Similar to [[resolveOperatorsUpWithPruning]], but does it top-down. */
+  def resolveOperatorsDownWithPruning(cond: TreePatternBits => Boolean,
+    ruleId: RuleId = UnknownRuleId)(rule: PartialFunction[LogicalPlan, LogicalPlan])
+  : LogicalPlan = {
+    if (!analyzed && cond.apply(self) && !isRuleIneffective(ruleId)) {
       AnalysisHelper.allowInvokingTransformsInAnalyzer {
```

Comment:
A bit of OCD, but do we really have to do this check for each recursive call? It seems a bit wasteful.

Reply:
Yeah, makes sense.
I'll see if I can remove allowInvokingTransformsInAnalyzer in a subsequent PR.

Reply:
Sounds good!